File: c++/leetcode/103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
  vector<vector<int>> zigzagLevelOrder(TreeNode *root)
  {
    vector<vector<int>> ans;

    if (!root)
      return ans;

    bool alt{};
    queue<TreeNode *> dq;
    dq.push(root);

    while (dq.size())
    {
      vector<int> vec;
      const int sz = dq.size();

      for (int i = 0; i < sz; ++i)
      {
        auto front = dq.front();
        vec.push_back(front->val);

        if (front->left)
          dq.push(front->left);

        if (front->right)
          dq.push(front->right);

        dq.pop();
      }

      if (alt)
        reverse(vec.begin(), vec.end());

      alt = !alt;
      ans.push_back(vec);
    }

    return ans;
  }
};

File: javascript/leetcode/103. Binary Tree Zigzag Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  const ans = [];

  if (!root) return ans;

  let alt = false;
  let q = [root];

  while (q.length) {
    const vec = [];
    const sz = q.length;

    for (let i = 0; i < sz; ++i) {
      const front = q[0];
      vec.push(front.val);

      if (front.left) q.push(front.left);

      if (front.right) q.push(front.right);

      q = q.slice(1);
    }

    if (alt) vec.reverse();

    alt = !alt;
    ans.push(vec);
  }

  return ans;
};
